src/api/actions.tsx

```diff
@@ -133,6 +141,85 @@ export const useIncidents = (): [IncidentsStateType, UseIncidentsActionType] =>
+      (async (pks: Incident["pk"][], ackBody: AcknowledgementBody) => {
+        dispatch(setOngoingBulkUpdate());
+        let acks: Acknowledgement[] = [];
+        for (const pk of pks) {
+          await api.postAck(pk, ackBody).then((ack: Acknowledgement) => {
+            acknowledgeIncident(pk);
+            acks.push(ack);
+          }).catch((error) => {
+            dispatch(unsetOngoingBulkUpdate());
+            throw new Error(error);
+          })
+        }
+        dispatch(unsetOngoingBulkUpdate());
+        return acks;
+      }),
+      [dispatch, acknowledgeIncident],
+  );
+
+    const bulkAddTicketUrlCallback = useCallback(
+        (async (pks: Incident["pk"][], ticketUrl: string) => {
+            dispatch(setOngoingBulkUpdate());
+            let responses: IncidentTicketUrlBody[] = [];
+            for (const pk of pks) {
+                await api.patchIncidentTicketUrl(pk, ticketUrl)
+                    .then((response: IncidentTicketUrlBody) => {
+                        addTicketUrl(pk, ticketUrl);
+                        responses.push(response);
+                    }).catch((error) => {
+                        dispatch(unsetOngoingBulkUpdate());
+                        throw new Error(error);
+                })
+            }
+            dispatch(unsetOngoingBulkUpdate());
+            return responses;
+        }),
+        [dispatch, addTicketUrl],
+    );
+
+    const bulkReopenIncidentsCallback = useCallback(
+        (async (pks: Incident["pk"][]) => {
+            dispatch(setOngoingBulkUpdate());
+            let reopenEvents: Event[] = [];
+            for (const pk of pks) {
+                await api.postIncidentReopenEvent(pk)
+                    .then((reopenEvent: Event) => {
+                        reopenIncident(pk);
+                        reopenEvents.push(reopenEvent);
+                    }).catch((error) => {
+                        dispatch(unsetOngoingBulkUpdate());
+                        throw new Error(error);
+                    })
+            }
+            dispatch(unsetOngoingBulkUpdate());
+            return reopenEvents;
+        }),
+        [dispatch, reopenIncident],
+    );
+
+    const bulkCloseIncidentsCallback = useCallback(
+        (async (pks: Incident["pk"][], description?: string) => {
+            dispatch(setOngoingBulkUpdate());
+            let closeEvents: Event[] = [];
+            for (const pk of pks) {
+                await api.postIncidentCloseEvent(pk, description)
+                    .then((closeEvent: Event) => {
+                        closeIncident(pk);
+                        closeEvents.push(closeEvent);
+                    }).catch((error) => {
+                        dispatch(unsetOngoingBulkUpdate());
+                        throw new Error(error);
+                    })
+            }
+            dispatch(unsetOngoingBulkUpdate());
+            return closeEvents;
+        }),
+        [dispatch, closeIncident],
+    );
+
```

Comment:
I have looked into it. The only areas for reuse here are _dispatch(setOngoingBulkUpdate())_ and _dispatch(unsetOngoingBulkUpdate())_. 
I could create a separate callback that calls _dispatch(setOngoingBulkUpdate())_, then calls a specific bulk operation callback, and then calls _dispatch(unsetOngoingBulkUpdate())_ (either on error, or on return from the bulk-op callback). But with TS, this new callback would require a **very** chunky type definition. This is because of bulk-ops callbacks all having different arguments and return values. This would make this part of the code much more convoluted. 
I think that repeating _dispatch(setOngoingBulkUpdate())_ and _dispatch(unsetOngoingBulkUpdate())_ is  actually a more clean and clear solution 🤔

Reply:
I don't think DRYing is worth it due to typescript. You can't really duck-type here.

Reply:
Okay, if its not viable then its fine